// unlog/src/unlog.ts

export enum LogLevel {
  LOG = 0,
  WARN,
  ERROR
};

const level_names = [
  "LOG",
  "WRN",
  "ERR"
];

var verbosity: LogLevel = LogLevel.LOG;

// This function only takes integer numbers as arguments.
function int_to_str(min_digits: number, value: number): string {
  if (value < 0) {
    return "-" + int_to_str(min_digits, -value);
  }
  
  if (value >= Math.pow(10, (min_digits - 1)) ||
      (value == 0 && min_digits <= 1)) {
    return "" + value;
  }

  return "0" + int_to_str(min_digits - 1, value);
}

function get_local_log_timestamp(): string {
  const date = new Date();
  const yrs = int_to_str(4, date.getFullYear());
  const mth = int_to_str(2, date.getMonth() + 1);
  const day = int_to_str(2, date.getDate());
  const hrs = int_to_str(2, date.getHours());
  const min = int_to_str(2, date.getMinutes());
  const sec = int_to_str(2, date.getSeconds());

  return `${yrs}-${mth}-${day} ${hrs}:${min}:${sec}`;
}

function get_utc_log_timestamp(): string {
  const date = new Date();
  const iso = date.toISOString();
  const date_string = iso.substr(0, 10);
  const time_string = iso.substr(11, 8);
  return `${date_string} ${time_string}`;
}

export function get_output(log_level: LogLevel, message: string): string {
  const ts = get_local_log_timestamp();
  const output = `${ts} ${level_names[log_level]} | ${message}`
  return output;
}

export function write(log_level: LogLevel, message: string): void {
  if (log_level < verbosity) {
    return;
  }
  const output = get_output(log_level, message);
  console.log(output);
}

export function set_verbosity(level: LogLevel) { verbosity = level; }

export function log(message: string) { write(LogLevel.LOG, message); }

export function warn(message: string) { write(LogLevel.WARN, message); }

export function error(message: string) { write(LogLevel.ERROR, message); }
