// unlog/src/unlog.js
"use strict";
exports.__esModule = true;
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["LOG"] = 0] = "LOG";
    LogLevel[LogLevel["WARN"] = 1] = "WARN";
    LogLevel[LogLevel["ERROR"] = 2] = "ERROR";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
;
var level_names = [
    "LOG",
    "WRN",
    "ERR"
];
var verbosity = LogLevel.LOG;
// This function only takes integer numbers as arguments.
function int_to_str(min_digits, value) {
    if (value < 0) {
        return "-" + int_to_str(min_digits, -value);
    }
    if (value >= Math.pow(10, (min_digits - 1)) ||
        (value == 0 && min_digits <= 1)) {
        return "" + value;
    }
    return "0" + int_to_str(min_digits - 1, value);
}
function get_local_log_timestamp() {
    var date = new Date();
    var yrs = int_to_str(4, date.getFullYear());
    var mth = int_to_str(2, date.getMonth() + 1);
    var day = int_to_str(2, date.getDate());
    var hrs = int_to_str(2, date.getHours());
    var min = int_to_str(2, date.getMinutes());
    var sec = int_to_str(2, date.getSeconds());
    return yrs + "-" + mth + "-" + day + " " + hrs + ":" + min + ":" + sec;
}
function get_utc_log_timestamp() {
    var date = new Date();
    var iso = date.toISOString();
    var date_string = iso.substr(0, 10);
    var time_string = iso.substr(11, 8);
    return date_string + " " + time_string;
}
function get_output(log_level, message) {
    var ts = get_local_log_timestamp();
    var output = ts + " " + level_names[log_level] + " | " + message;
    return output;
}
exports.get_output = get_output;
function write(log_level, message) {
    if (log_level < verbosity) {
        return;
    }
    var output = get_output(log_level, message);
    console.log(output);
}
exports.write = write;
function set_verbosity(level) { verbosity = level; }
exports.set_verbosity = set_verbosity;
function log(message) { write(LogLevel.LOG, message); }
exports.log = log;
function warn(message) { write(LogLevel.WARN, message); }
exports.warn = warn;
function error(message) { write(LogLevel.ERROR, message); }
exports.error = error;
